File: ruby/sinatra/spec/controllers/users/registrations_controller_spec.rb
RSpec.describe Users::RegistrationsController, type: :request do
  describe "GET /current_user" do
    context "when a User is authenticated" do
      let(:user) { create(:user) }

      before do
        get "/current_user", nil, { "HTTP_AUTHORIZATION" => user.jwt }
      end

      it "returns a JSON object" do
        expect(last_response.body).to be_a String
        expect(JSON.parse(last_response.body)).to have_key "id"
      end

      it "returns the instance of User" do
        data = JSON.parse(last_response.body)
        expect(data["id"]).to eq user.id
      end

      it "returns a ok HTTP status" do
        expect(last_response.status).to be 200
      end
    end

    context "when a User is not authenticated" do
      it "returns a unauthorized HTTP status" do
        get "/current_user"
        expect(last_response.status).to be 401
      end
    end
  end

  describe "POST /users" do
    let(:email) { "user@example.com" }
    let(:password) { "password" }

    context "with proper params" do
      before do
        post "/users", params: { user: { email:, password:, password_confirmation: password } }
      end

      it "returns a JSON object" do
        expect(last_response.body).to be_a String
        expect(JSON.parse(last_response.body)).to have_key "id"
      end

      it "creates an instance of User" do
        expect(User.count).to eq 1
      end

      it "returns the new instance of User" do
        data = JSON.parse(last_response.body)
        expect(data["email"]).to eq email
      end

      it "returns a created HTTP status" do
        expect(last_response.status).to be 201
      end

      it "returns a Authorization header" do
        expect(last_response["Authorization"]).to eq User.first.jwt
      end
    end

    context "without proper params" do
      before do
        post "/users", params: { user: { email: nil, password: nil } }
      end

      it "returns a JSON object" do
        expect(last_response.body).to be_a String
        expect(JSON.parse(last_response.body)).to have_key "errors"
      end

      it "does not create an instance of User" do
        expect(User.count).to eq 0
      end

      it "returns a list of error messages" do
        data = JSON.parse(last_response.body)
        expect(data["errors"]).to contain_exactly("Email can't be blank", "Password can't be blank")
      end

      it "returns a HTTP status of unprocessable_entity" do
        expect(last_response.status).to be 422
      end
    end
  end
end

File: node/express/spec/controllers/users/registrations.controller.spec.ts
import supertest from "supertest"
import { beforeEach } from "vitest"
import app from "src/app.ts"
import prisma, { type User } from "src/models/user.model.ts"

describe("UsersRegistrationsController", () => {
  let response: supertest.Response
  const request = supertest(app)
  const email = "user@example.com"
  const password = "password"

  const createUser = async () => await prisma.user.create({ data: { email, password } })

  describe("GET /current_user", () => {
    describe("when a User is authenticated", () => {
      let user: User

      beforeEach(async () => {
        user = await createUser()
        response = await request.get("/current_user").set("Authorization", user.jwt)
      })

      it("returns a JSON object", () => {
        expect(response.type).toBe("application/json")
      })

      it("returns the instance of User", () => {
        expect(response.body).toMatchObject({ id: user.id })
      })

      it("returns a ok HTTP status", () => {
        expect(response.status).toBe(200)
      })
    })

    describe("when a User is not authenticated", () => {
      it("returns a unauthorized HTTP status", async () => {
        response = await request.get("/current_user")
        expect(response.status).toBe(401)
      })
    })
  })

  describe("POST /users", () => {
    describe("with proper params", () => {
      beforeEach(async () => {
        response = await request.post("/users").send({ user: { email, password } })
      })

      it("returns a JSON object", () => {
        expect(response.type).toBe("application/json")
      })

      it("creates an instance of User", async () => {
        await expect(prisma.user.count()).resolves.toBe(1)
      })

      it("returns the new instance of User", () => {
        expect(response.body).toMatchObject({ email })
      })

      it("returns a created HTTP status", () => {
        expect(response.status).toBe(201)
      })

      it("returns a Authorization header", async () => {
        const user = await prisma.user.findFirst()
        expect(response.headers.authorization).toBeDefined()
        expect(response.headers.authorization).toBe(user?.jwt)
      })
    })

    describe("without proper params", () => {
      beforeEach(async () => {
        response = await request.post("/users").send({ user: { email: null, password: null } })
      })

      it("returns a JSON object", () => {
        expect(response.type).toBe("application/json")
      })

      it("does not create an instance of User", async () => {
        await expect(prisma.user.count()).resolves.toBe(0)
      })

      it("returns a list of error messages", () => {
        expect(response.body).toStrictEqual({
          errors: { email: "Email can't be blank", password: "Password can't be blank" }
        })
      })

      it("returns a unprocessable_entity HTTP status", () => {
        expect(response.status).toBe(422)
      })
    })
  })
})
